cocotb/share/lib/simulator/simulatormodule.c

```diff
@@ -46,6 +46,161 @@ struct module_state {
     PyObject *error;
 };
 
+
+/* define the extension types */
+
+/* for gpi_sim_hdl */
+static PyTypeObject gpi_sim_hdl_Type;
+
+typedef struct {
+    PyObject_HEAD
+    gpi_sim_hdl hdl;
+} gpi_sim_hdl_Object;
+
+static PyObject *gpi_sim_hdl_repr(gpi_sim_hdl_Object *self) {
+    PyTypeObject *type = Py_TYPE(self);
+    return PyUnicode_FromFormat("<%s at %p>",
+                                  type->tp_name, self->hdl);
+}
+static Py_hash_t gpi_sim_hdl_hash(gpi_sim_hdl_Object *self) {
+    Py_hash_t ret = (Py_hash_t)self->hdl;
+    if (ret == (Py_hash_t)-1) {
+        ret = (Py_hash_t)-2;
+    }
+    return ret;
+}
+
+static PyObject *gpi_sim_hdl_richcompare(PyObject *self, PyObject *other, int op) {
+    gpi_sim_hdl_Object *self_hdl_obj;
+    gpi_sim_hdl_Object *other_hdl_obj;
+
+    if (
+            Py_TYPE(self) != &gpi_sim_hdl_Type ||
+            Py_TYPE(other) != &gpi_sim_hdl_Type) {
+        Py_RETURN_NOTIMPLEMENTED;
+    }
+
+    self_hdl_obj = (gpi_sim_hdl_Object *)self;
+    other_hdl_obj = (gpi_sim_hdl_Object *)other;
+
+    switch (op) {
+        case Py_EQ: return PyBool_FromLong(self_hdl_obj->hdl == other_hdl_obj->hdl);
+        case Py_NE: return PyBool_FromLong(self_hdl_obj->hdl != other_hdl_obj->hdl);
+        default: Py_RETURN_NOTIMPLEMENTED;
+    }
+}
+
+static PyTypeObject gpi_sim_hdl_Type = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    .tp_name = "simulator.gpi_sim_hdl",
+    .tp_basicsize = sizeof(gpi_sim_hdl_Object),
+    .tp_flags = Py_TPFLAGS_DEFAULT,
+    .tp_repr = (reprfunc)gpi_sim_hdl_repr,
+    .tp_hash = (hashfunc)gpi_sim_hdl_hash,
+    .tp_richcompare = gpi_sim_hdl_richcompare,
+};
+
+static PyObject *gpi_sim_hdl_New(gpi_sim_hdl hdl) {
+    gpi_sim_hdl_Object *obj;
+    if (hdl == NULL) {
+        Py_RETURN_NONE;
+    }
+    obj = PyObject_New(gpi_sim_hdl_Object, &gpi_sim_hdl_Type);
+    if (obj == NULL) {
+        return NULL;
+    }
+    obj->hdl = hdl;
+    return (PyObject *)obj;
+}
```

Comment:
Note that #1442 would allow this chunk of code to become a template used for all three types, rather than it needing to be repeated 2 more times below